Assets/HoloToolkit/UX/Scripts/BoundingBoxes/BoundingBox.cs

```diff
@@ -0,0 +1,482 @@
+﻿//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+//
+using System;
+
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace HoloToolkit.Unity.UX
+{
+    /// <summary>
+    /// Base class for bounding box objects
+    /// </summary>
+    public class BoundingBox : MonoBehaviour
+    {
+        public enum BoundsCalculationMethodEnum
+        {
+            MeshFilterBounds,   // Better for flattened objects - this mode also treats RectTransforms as quad meshes
+            RendererBounds,     // Better for objects with non-mesh renderers
+            Colliders,          // Better if you want precise control
+            Default,            // Use the default method (RendererBounds)
+        }
+
+        public enum FlattenModeEnum
+        {
+            DoNotFlatten,   // Always use XYZ axis
+            FlattenX,       // Flatten the X axis
+            FlattenY,       // Flatten the Y axis
+            FlattenZ,       // Flatten the Z axis
+            FlattenAuto,    // Flatten the smallest relative axis if it falls below threshold
+        }
+
+        #region public
+
+        public Action OnFlattenedAxisChange;
+
+        public virtual FlattenModeEnum FlattenPreference {
+            get { return flattenPreference; }
+            set { flattenPreference = value; }
+        }
+
+        public virtual BoundsCalculationMethodEnum BoundsCalculationMethod {
+            get { return boundsCalculationMethod; }
+            set { boundsCalculationMethod = value; }
+        }
+
+        /// <summary>
+        /// The target object being manipulated
+        /// </summary>
+        public virtual GameObject Target
+        {
+            get
+            {
+                return target;
+            }
+            set
+            {
+                if (target != value)
+                {
+                    // Send a message to the new / old targets
+                    if (value != null)
+                    {
+                        value.SendMessage("OnTargetSelected", SendMessageOptions.DontRequireReceiver);
+                    }
+                    if (target != null)
+                    {
+                        target.SendMessage("OnTargetDeselected", SendMessageOptions.DontRequireReceiver);
```

Comment:
Why are we using `SendMessage`?  We can probably find a more elegant solution to this that doesn't rely on a string reference, and will be able to have a symbolic reference instead.

Reply:
Agreed, I don't like SendMessage. Any suggestions? Should we just let OnTargetDeselected be an opt-in Action?

Reply:
We could add an event for the Input Manager to raise and only this class listens to.

Reply:
Not sure how I feel about bringing bounding box events into InputManager since their purpose is so specific.

Some kind of ManipulationManager of BoundingBoxManager with TargetSelected / TargetDeselected events might be better. Sort of like FocusManager.

I'll remove these events for now since they're not crucial.